### DISCUSSION TITLE
Advice re ESTOP, FHOLD, CS_RES = GRBL/CNCShield Abort, Hold, Resume

### DISCUSSION THREAD

**Discussion by hmmueller** [Paciente8159/uCNC#721] [Q&A] [answered]
First, I am not sure I understand the difference between ESTOP and FHOLD: Both, if connected to GND, will stop everything. When then pulled up to +5V, I assume with FHOLD everything starts to move; but with ESTOP it wont. What then is the separate "Resume"?
... or are all three thought to be shortly closed switches - but how would I then wire FHOLD + CS_RES e.g. with a hood switch?

I have a NC+NO emergency button; the NC side cuts out the 230V power (to steppers and spindle and coolant pump) if I hit it. I assume I connect the NO side (which closes when I hit the emergency button) to ESTOP (PC0) and GND (which are currently jumpered, as is FHOLD).

Last problem: My cheap Chinese emergency button does not permanently close the NO contact when hit - it closes it for a moment, but then (due to mechanical sloppiness) opens it again. Is that fine with the ESTOP pin, or should I make sure that the NO really and permanently closes when the button is hit?

Thanks for any enlightenment (which I did not find by searching various GRBL forums on the internet ...).
H.M.

**Comment by Paciente8159** [comment, marked as answer]
> First, I am not sure I understand the difference between ESTOP and FHOLD: Both, if connected to GND, will stop everything. When then pulled up to +5V, I assume with FHOLD everything starts to move; but with ESTOP it wont. What then is the separate "Resume"?
> ... or are all three thought to be shortly closed switches - but how would I then wire FHOLD + CS_RES e.g. with a hood switch?

ESTOP - stands for emergency stop. Triggering this input will kill all motions abruptly. The modal states will be reset and your machine will lose track of it's position. The current working tool will also be stopped.

FHOLD - stands for feed hold is just a pause button. The motors will come to a controlled pause and tool will stop. Machine will keep all it's internal states and position. You can use CS_RES (Cycle start - resume) to continue were you left. The tool will be started. A programmable pause time can occur to allow the tool to get up to speed and then motion will restart.


> I have a NC+NO emergency button; the NC side cuts out the 230V power (to steppers and spindle and coolant pump) if I hit it. I assume I connect the NO side (which closes when I hit the emergency button) to ESTOP (PC0) and GND (which are currently jumpered, as is FHOLD).

Super bad idea. If the common terminal to that NC and NO terminals is the same you will be injecting 230V on your board.
Using a 230V kill switch will have a similar (more effective in some way) effect to the ESTOP pin. You can free the ESTOP pin for other purposes in that situation.


> Last problem: My cheap Chinese emergency button does not permanently close the NO contact when hit - it closes it for a moment, but then (due to mechanical sloppiness) opens it again. Is that fine with the ESTOP pin, or should I make sure that the NO really and permanently closes when the button is hit?

That is not a problem with ESTOP. As soon as it is triggered it will go into Alarm mode and shutdown everything. Even if the switch does not keep the contact opened/closed. The difference is that if the emergency switch is not pressed you will be able to send a soft reset command to get out of alarm mode.

**Reply by hmmueller** [reply]
Thanks, as always, for quick answers.

Re FHOLD and CS_RES, just to be sure: After triggering FHOLD (i.e. opening it), again closing it will NOT resume anything - is that right? So, if I add a hood with a switch that opens 

- when I lift the hood, motors will stop  
- when I then close the hood, nothing will happen, until ...
- ... I open (or close?) some "resume button" that's wired to CS_RES.

Re NC+NO :-) That would be bad indeed. There's, of course, no common terminal - green (NO) and orange (NC) side are far away from each other:

![EmergencyStop](https://github.com/user-attachments/assets/c36768e9-05d0-4a0b-a356-6645b5dee2f9)

Re ESTOP - ok. But I wont wire the ESTOP, as I "hard kill" motor power (spindle as well as stepper power) at the 230 V level.

H.M.

**Reply by Paciente8159** [reply]
When you press FHOLD you will send a HOLD command. You don't need to keep pressing the FHOLD forever. A pulse signal will trigger FHOLD. The machine will stay in Hold until you tell it to resume.

But notice this. If you let CS_RES unwired or fluctuating you might get an accidental resume trigger from environmental noise.

> Re NC+NO :-) That would be bad indeed. There's, of course, no common terminal - green (NO) and orange (NC) side are far away from each other:

If they a electrically independent buttons then yes you can use them both..but I don't recommend it.
Using it as a general power cut would be the best bet securely shutdown all operations. Carefull though. If this is not powering off the board/PC then you need to make sure the controller board is stopped before reenable spindle and steppers. The board might be still executing code and commanding stuff because it's not aware of the emergency stop.

**Reply by hmmueller** [reply]
All understood. Your last point, I'd say, (make sure that controller does not emit commands after an emergency stop) can be handled in 3 different ways:

a) Use the NO contact on the emergency button to "Abort" (in GRBL talk) the controller. You don't recommend it, I don't see why - mechanical coupling of both effects (no more power to motors; stopping controller) is, safety-wise, the way to go, I'd assume.
b) Use a relay (or the like) in the power lines which, on no power voltage, "Aborts" the controller. A little more effort, potentially a failing device (the relay).
c) Don't tell the controller about the Abort - "make sure" is a manual process. This will backfire one day, I'd say :-)

I think I'll go with a) - cheapest -, until I get a problem. Then I'll have learned something.

Thanks!
H.M.

**Reply by Paciente8159** [reply]
I don't recommend the NO contact because in case of wire failure or you forget to  connect it it will never be detected.

The NC gives you that advantage.
